geode-core/src/distributedTest/java/org/apache/geode/internal/cache/versions/TombstoneDUnitTest.java

```diff
@@ -117,35 +118,74 @@ public void testTombstoneGcMessagesBetweenPersistentAndNonPersistentRegion() {
     });
   }
 
+
   @Test
-  public void testGetOldestTombstoneTimeReplicate() {
+  public void testWhenAnOutOfRangeTimeStampIsSeenWeExpireItInReplicateTombstoneSweeper() {
     VM server1 = VM.getVM(0);
     VM server2 = VM.getVM(1);
+    final int FAR_INTO_THE_FUTURE = 1000000; // 1 million millis into the future
 
+    final int count = 10;
     server1.invoke(() -> {
-      createCacheAndRegion(REPLICATE_PERSISTENT);
-      region.put("K1", "V1");
-      region.put("K2", "V2");
+      createCacheAndRegion(RegionShortcut.REPLICATE_PERSISTENT);
+      for (int i = 0; i < count; i++) {
+        region.put("K" + i, "V" + i);
+      }
     });
 
-    server2.invoke(() -> createCacheAndRegion(REPLICATE));
+    server2.invoke(() -> createCacheAndRegion(RegionShortcut.REPLICATE));
 
     server1.invoke(() -> {
+      TombstoneService.TombstoneSweeper tombstoneSweeper =
+          ((InternalCache) cache).getTombstoneService().getSweeper((LocalRegion) region);
+
+      RegionEntry regionEntry = ((LocalRegion) region).getRegionEntry("K0");
+      VersionTag<?> versionTag = regionEntry.getVersionStamp().asVersionTag();
+      versionTag.setVersionTimeStamp(System.currentTimeMillis() + FAR_INTO_THE_FUTURE);
+
+      TombstoneService.Tombstone modifiedTombstone =
+          new TombstoneService.Tombstone(regionEntry, (LocalRegion) region,
+              versionTag);
+      tombstoneSweeper.tombstones.add(modifiedTombstone);
+      tombstoneSweeper.checkOldestUnexpired(System.currentTimeMillis());
       // Send tombstone gc message to vm1.
-      region.destroy("K1");
+      assertThat(tombstoneSweeper.getOldestTombstoneTime()).isEqualTo(0);
+    });
+  }
 
+  @Test
+  public void testGetOldestTombstoneTimeForReplicateTombstoneSweeper() {
+    VM server1 = VM.getVM(0);
+    VM server2 = VM.getVM(1);
+    final int count = 10;
+    server1.invoke(() -> {
+      createCacheAndRegion(RegionShortcut.REPLICATE_PERSISTENT);
+      for (int i = 0; i < count; i++) {
+        region.put("K" + i, "V" + i);
+      }
+    });
+
+    server2.invoke(() -> createCacheAndRegion(RegionShortcut.REPLICATE));
+
+    server1.invoke(() -> {
       TombstoneService.TombstoneSweeper tombstoneSweeper =
           ((InternalCache) cache).getTombstoneService().getSweeper((LocalRegion) region);
+      // Send tombstone gc message to vm1.
+      for (int i = 0; i < count; i++) {
+        region.destroy("K" + i);
+        assertThat(
```

Comment:
I am not sure I understand the purpose of this assert. Even without the fix, the assert will succeed.

Reply:
We are getting the remaining time until the tombstone should be expired. We are expecting it to be greater than 0.  Hence the assert. Then we force a GC and we expect it to be 0. Effectively this test is validating that performing a gc clears out the tombstones.

Reply:
First, usually immediately after `Region.destroy()`, the difference between `tombstoneSweeper.getOldestTombstoneTime()` and `System.currentTimeMillis()` is only a few milliseconds. So asserting that 
```
tombstoneSweeper.getOldestTombstoneTime()
                + TombstoneService.REPLICATE_TOMBSTONE_TIMEOUT_DEFAULT - System.currentTimeMillis())
                    .isGreaterThan(0)
```
is like asserting `REPLICATE_TOMBSTONE_TIMEOUT_DEFAULT` is greater than 0, i.e. asserting 600000 > 0. There is nothing wrong with that. However, a better assertion would be the one you deleted:
```
      assertThat(tombstoneSweeper.getOldestTombstoneTime()).isGreaterThan(0)
          .isLessThan(((InternalCache) cache).cacheTimeMillis());
```
Second, if we expect that GC clears out tombstones, I would recommend moving the following assertion
```
assertThat(tombstoneSweeper.getOldestTombstoneTime()).isEqualTo(0)
```
 into the `for` loop, instead of after the `for` loop.
Please correct me if I am wrong.

Reply:
This would be wrong cacheTimeMillis has no guarantees about the next milliseconds value.  So dealing with values that small could cause a problem. Further,  we want to validate that we have an unexpired tombstone. What was there is doesn't account for the extended validity period that is the EXPIRY_TIMEOUT.

Reply:
For DUnit test, `cacheTimeMillis()` is returns the same value as `System.currentTimeMillis()`.  To validate an unexpired tombstone, besides asserting getOldestTombstoneTime()  + REPLICATE_TOMBSTONE_TIMEOUT_DEFAULT - now > 0, there is also another condition msUntilTombstoneExpires < EXPIRY_TIME, as suggested in your code change in `hasExpired()`. Anyway, what I have suggested are for improvements.